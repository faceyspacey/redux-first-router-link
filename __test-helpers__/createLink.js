import React from 'react'
import renderer from 'react-test-renderer'
import { createStore, applyMiddleware, compose } from 'redux'
import { Provider } from 'react-redux'
import { connectRoutes } from 'redux-first-router'

import Link from '../src/Link'
import NavLink from '../src/NavLink'

const createLink = (props, initialPath, options) => {
  const isNavLink = !!initialPath
  const link = isNavLink ? <NavLink {...props} /> : <Link {...props} />

  const routesMap = {
    FIRST: '/first',
    SECOND: '/second/:param'
  }

  const { middleware, enhancer, reducer } = connectRoutes(routesMap, {
    initialEntries: [initialPath || '/'],
    initialIndex: 0,
    keyLength: 6,
    ...options
  })

  const middlewares = applyMiddleware(middleware)
  const enhancers = compose(
    enhancer,
    middlewares
  )
  const rootReducer = (state = {}, action = {}) => ({
    location: reducer(state.location, action)
  })

  const store = createStore(rootReducer, enhancers)
  const component = renderer.create(
    <Provider store={store}>{link}</Provider>,
    { createNodeMock: () => 'node' },
  )

  return {
    component,
    tree: component.toJSON(),
    store
  }
}

export default createLink
export const createNavLink = (path, props) => createLink(props, path)

export const event = { preventDefault: () => undefined, button: 0 }
