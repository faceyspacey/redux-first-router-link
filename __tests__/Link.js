import { NOT_FOUND, addRoutes } from 'redux-first-router'
import createLink, { event } from '../__test-helpers__/createLink'

test('ON_CLICK: dispatches location-aware action', () => {
  const { tree, store } = createLink({
    to: '/first',
    children: 'CLICK ME'
  }) /*? $.tree */

  expect(tree).toMatchSnapshot() /*? store.getState() */

  tree.props.onClick(event)

  const { location } = store.getState() /*? $.location */

  expect(location.pathname).toEqual('/first')
  expect(location.type).toEqual('FIRST')
})

it('ON_CLICK: does NOT dispatch if shouldDispatch === false', () => {
  const { tree, store } = createLink({ to: '/first', shouldDispatch: false })

  expect(tree).toMatchSnapshot()

  tree.props.onClick(event)

  const { location } = store.getState() /*? $.location */

  expect(location.pathname).not.toEqual('/first')
  expect(location.type).not.toEqual('FIRST')
  expect(location.pathname).toEqual('/')
})

it('ON_CLICK: does NOT dispatch if onClick returns false', () => {
  const { tree, store } = createLink({ to: '/first', onClick: () => false })

  expect(tree).toMatchSnapshot()

  tree.props.onClick(event)

  const { location } = store.getState() /*? $.location */

  expect(location.pathname).not.toEqual('/first')
  expect(location.type).not.toEqual('FIRST')
  expect(location.pathname).toEqual('/')
})

it('ON_CLICK: DOES dispatch if onClick returns true', () => {
  const { tree, store } = createLink({ to: '/first', onClick: () => true })

  expect(tree).toMatchSnapshot()

  tree.props.onClick(event)

  const { location } = store.getState() /*? $.location */

  expect(location.pathname).toEqual('/first')
  expect(location.type).toEqual('FIRST')
})

it('ON_CLICK: DOES dispatch if onClick returns undefined', () => {
  const { tree, store } = createLink({
    to: '/first',
    onClick: () => undefined
  })

  expect(tree).toMatchSnapshot()

  tree.props.onClick(event)

  const { location } = store.getState() /*? $.location */

  expect(location.pathname).toEqual('/first')
  expect(location.type).toEqual('FIRST')
})

it('ON_MOUSE_DOWN: dispatches action onMouseDown if down === true', () => {
  const { tree, store } = createLink({
    to: '/first',
    down: true
  }) /*? $.tree */

  expect(tree).toMatchSnapshot()

  tree.props.onClick(event) // e.preventDefault() called and nothing dispatched + linking prevented

  let { location } = store.getState()
  expect(location.pathname).toEqual('/')

  tree.props.onMouseDown(event)

  location = store.getState().location /*? */

  expect(location.pathname).toEqual('/first')
  expect(location.type).toEqual('FIRST')
})

it('ON_CLICK: dispatches redirect if redirect === true', () => {
  const { tree, store } = createLink({
    to: '/first',
    children: 'CLICK ME',
    redirect: true
  }) /*? $.tree */

  expect(tree).toMatchSnapshot() /*? store.getState() */

  tree.props.onClick(event)

  const { location } = store.getState() /*? $.location */

  expect(location.pathname).toEqual('/first')
  expect(location.kind).toEqual('redirect')
})

it('ON_TOUCH_START: dispatches action onTouchStart if down === true', () => {
  const { tree, store } = createLink({
    to: '/first',
    down: true
  }) /*? $.tree */

  expect(tree).toMatchSnapshot()

  tree.props.onTouchStart(event)

  const { location } = store.getState() /*? $.location */

  expect(location.pathname).toEqual('/first')
  expect(location.type).toEqual('FIRST')
})

it('converts href as array of strings to path', () => {
  const { tree, store } = createLink({ to: ['second', 'bar'] }) /*? $.tree */

  expect(tree).toMatchSnapshot() /*? store.getState() */

  tree.props.onClick(event)

  const { location } = store.getState() /*? $.location */

  expect(location.pathname).toEqual('/second/bar')
  expect(location.type).toEqual('SECOND')
})

it('converts href as action object to path', () => {
  const action = { type: 'SECOND', payload: { param: 'bar' } }
  const { tree, store } = createLink({ to: action }) /*? $.tree */

  expect(tree).toMatchSnapshot() /*? store.getState() */

  tree.props.onClick(event)

  const { location } = store.getState() /*? $.location */

  expect(location.pathname).toEqual('/second/bar')
  expect(location.type).toEqual('SECOND')
})

it('converts href as non-matched action object to "#" for path', () => {
  const action = { type: 'MISSED' }
  const { tree, store } = createLink({ to: action }) /*? $.tree */

  expect(tree.props.href).toEqual('#')
  expect(tree).toMatchSnapshot()

  tree.props.onClick(event)

  const { location } = store.getState() /*? $.location */
  expect(location.type).toEqual(NOT_FOUND)
})

it('converts invalid href to "#" for path', () => {
  const { tree, store } = createLink({ to: '' }) /*? $.tree */

  expect(tree.props.href).toEqual('#')
  expect(tree).toMatchSnapshot()

  tree.props.onClick(event)

  const { location } = store.getState() /*? $.location */
  expect(location.type).toEqual(NOT_FOUND)
})

it('supports custom HTML tag name', () => {
  const { tree, store } = createLink({
    to: 'somewhere',
    tagName: 'div'
  }) /*? $.tree */

  expect(tree).toMatchSnapshot()
})

it('supports custom HTML tag name which is still a link', () => {
  const { tree, store } = createLink({
    to: 'somewhere',
    tagName: 'a'
  }) /*? $.tree */

  expect(tree).toMatchSnapshot()
})

test('with basename options generates url with basename', () => {
  const options = { basename: '/base-foo' }
  const { tree, store } = createLink(
    {
      to: '/first',
      children: 'CLICK ME'
    },
    '/',
    options
  )

  expect(tree.props.href).toEqual('/base-foo/first')
})

test('rerenders href when route url is updated', () => {
  const { component, store } = createLink({ to: { type: 'FIRST' } })
  expect(component.toJSON().props.href).toEqual('/first')
  store.dispatch(addRoutes({ FIRST: '/updatedfirst' }))
  expect(component.toJSON().props.href).toEqual('/updatedfirst')
})
