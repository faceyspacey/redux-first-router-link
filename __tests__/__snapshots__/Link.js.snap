// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ON_CLICK: DOES dispatch if onClick returns true 1`] = `
<a
  href="/first"
  onClick={[Function]}
/>
`;

exports[`ON_CLICK: DOES dispatch if onClick returns undefined 1`] = `
<a
  href="/first"
  onClick={[Function]}
/>
`;

exports[`ON_CLICK: dispatches location-aware action 1`] = `
<a
  href="/first"
  onClick={[Function]}
>
  CLICK ME
</a>
`;

exports[`ON_CLICK: dispatches redirect if redirect === true 1`] = `
<a
  href="/first"
  onClick={[Function]}
>
  CLICK ME
</a>
`;

exports[`ON_CLICK: does NOT dispatch if onClick returns false 1`] = `
<a
  href="/first"
  onClick={[Function]}
/>
`;

exports[`ON_CLICK: does NOT dispatch if shouldDispatch === false 1`] = `
<a
  href="/first"
  onClick={[Function]}
/>
`;

exports[`ON_MOUSE_DOWN: dispatches action onMouseDown if down === true 1`] = `
<a
  href="/first"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
/>
`;

exports[`ON_TOUCH_START: dispatches action onTouchStart if down === true 1`] = `
<a
  href="/first"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
/>
`;

exports[`converts NOT_FOUND to /not_found 1`] = `
<a
  href="/not-found"
  onClick={[Function]}
/>
`;

exports[`converts href as action object to path 1`] = `
<a
  href="/second/bar"
  onClick={[Function]}
/>
`;

exports[`converts href as array of strings to path 1`] = `
<a
  href="/second/bar"
  onClick={[Function]}
/>
`;

exports[`converts href as non-matched action object to "#" for path 1`] = `
<a
  href="#"
  onClick={[Function]}
/>
`;

exports[`converts invalid href to "#" for path 1`] = `
<a
  href="#"
  onClick={[Function]}
/>
`;

exports[`supports custom HTML tag name 1`] = `
<div
  onClick={[Function]}
/>
`;

exports[`supports custom HTML tag name which is still a link 1`] = `
<a
  href="somewhere"
  onClick={[Function]}
/>
`;
